stm32: QSPI NOR flash controller support

erwango opened the pull request:
Add dma based QSPI NOR flash controller for stm32l4 and stm32wb series.

Driver configures both NOR flash and also QSPI hardware block.
Forked from https://github.com/zephyrproject-rtos/zephyr/pull/25806 with addition of optional DMA support and reworked to make use jesd216 library.

Uses https://github.com/zephyrproject-rtos/zephyr/pull/31303

mnkp committed:
drivers/clock_control: stm32: Add AHB3 bus support

AHB3 bus support is added for compatible series.
Additionaly, fix condition for AHB2 support and fix
formatting

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

erwango committed:
dts/bindings: qspi: Add stm32 qspi binding

qspi bus controller bindings for stm32 family.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts/bindings: flash_controller: Add stm32 qspi-nor flash controller

Add a binding to describe stm32 qspi-nor flash controller

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/dma: stm32: Preparation for QSPI DMA mode support

In preparation for QSPI DMA mode:
-Add a possibility to override driver by the HAL DMA. In that case
stream is set as busy and no configuration nor treatment is done.
In case of interrupt, flags clearing is let to HAL.
-Treat Half Transfer interrupt prior to Transfer Complete for the
cases were both IRQ are both raised at the time IRQ handler is called

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
dts: Add stm32 quadspi support on L4 series

Add device tree description for qspi node on STM32L4 and
signals on whole family.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

erwango committed:
boards: disco_l475_iot1: Enable support for MX25R6435F QSPI NOR

Configure QSPI NOR support and MX25R6435F on disco_l475_iot1 board.
Set MX25R6435F as flash controller and arrange partitions to take
newlay available space into account.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
^^ @giansta, @r2r0

erwango commented:
^^@mtahirbutt @hjuul

pabigot commented on drivers/flash/spi_nor.h:
```diff
@@ -24,6 +24,7 @@
 #define SPI_NOR_CMD_PP          0x02    /* Page program */
 #define SPI_NOR_CMD_SE          0x20    /* Sector erase */
 #define SPI_NOR_CMD_BE_32K      0x52    /* Block erase 32KB */
+#define SPI_NOR_CMD_RSFDP       0x5A    /* Read SFDP */
```
Could you use the existing `JESD216_CMD_READ_SFDP` instead of adding this here?

erwango replied:
Sure I will

pabigot reviewed (commented):
Preliminary comments.  I hope to have hardware to test this within the next two weeks.

The spi-nor sample and probably at least one flash subsystem sample should be updated to support a board  (preferably B-L475E-IOT01A since that's what I've got coming...).

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -252,6 +252,18 @@ DMA_STM32_EXPORT_API int dma_stm32_configure(const struct device *dev,
 	id = id - STREAM_OFFSET;
 
+	/* Check potential DMA override */
+	if (config->linked_channel == 0x7F) {
```
Yeah. Tbh, I'm not quite sure where to put a #define. I'm violating DMA API to make it work together with DMA HAL api (cf dedicated commit), so I don't intend to update it for my purpose (unless one thinks it is legitimate, but I have doubts myself).
One solution would be to add a dma_stm32 specific include.

pabigot replied:
I'd just like to see a named identifier as a clue to what that 0x7F signifies; putting the define into this source file is fine from my perspective.

erwango commented on drivers/dma/dma_stm32.c:
```diff
@@ -97,14 +97,15 @@ static void dma_stm32_irq_handler(const struct device *dev, uint32_t id)
 		stream->dma_callback(dev, stream->user_data, callback_arg, 0);
-	} else if (dma_stm32_is_ht_active(dma, id)) {
-		dma_stm32_clear_ht(dma, id);
+		dma_stm32_clear_tc(dma, id);
```
The other way round didn't work with my purpose as DMA HAL needs flags to be still available when the callback is called to work as expected.
This being said, I might need to test more thoroughly other cases to be sure I didn't break anything and use this behavior only in case of "DMA channel override"

pabigot replied:
is this still an issue?

erwango replied:
> is this still an issue?

For sure I can't make it work the other way round due to use of HAL API. So this is still needed, yes.

pabigot replied:
I meant, did you still need to test more thoroughly, or are you satisfied that we can ignore the change?

erwango replied:
It was tested as much as I could and no issue was find. Anyway, I just protected this behavior so we should be safe on that front now.

erwango commented on dts/bindings/qspi/st,stm32-qspi.yaml:
```diff
@@ -0,0 +1,40 @@
+      required: true
+
+    dma-names:
+      required: true
```
You're totally right, I'll add a description to state that "fred and ethel" is not a valid name (and few more details ;-))

erwango replied:
Type are inherited, but I updated the descriptions.

erwango commented:
Txs @pabigot for the initial review. Will take those into account.

> The spi-nor sample and probably at least one flash subsystem sample should be updated to support a board (preferably B-L475E-IOT01A since that's what I've got coming...).

I agree that we need a way to test this driver in CI (or at least compile it). I think I can find a way to tweak `samples/drivers/flash_shell` to achieve this.
Though, I'm not sure about the spi-nor sample, do you mean `samples/drivers/jesd216` ?

gmarull commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -0,0 +1,927 @@
+
+	/* Initialize DMA HAL */
+	__HAL_LINKDMA(&dev_data->hqspi, hdma, hdma);
+	HAL_DMA_Init(&hdma);
```
As far as I understand, Zephyr DMA API is not used here. What is the reason? Other STM32 drivers such as SPI don't do that, I wonder if the way non-LL HAL is built forces this choice.

erwango replied:
That's indeed the point.
DMA calls are done inside QSPI HAL functions, it not possible to unplug. And there is no QSPI LL API.
Alternative is to go with a full CMSIS port of the driver, but we'll lose abstraction capability (and would require significant dev and re-validation effort I can't afford on my side).
Current solution is a bit frustrating, but I think this is the best available option.

gmarull replied:
@erwango got it. I think it's better to make more STM32 hw capabilities available before optimizing, so I think it's ok for now 👍

hjuul commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -0,0 +1,927 @@
+	/* TODO: This init is limited to G0/G4/L0/L4/L5/WB */
+	hdma.Init.PeriphDataAlignment = table_p_size[index];
+	hdma.Init.MemDataAlignment = table_m_size[index];
+	hdma.Init.Request = dma_cfg.dma_slot;
```
I was trying to make this work for stm32f7, but get a compile error here when building `samples/drivers/flash_shell`.

If you could extend support to `DMA_STM32_V1` also I can help to verify it on _stm32f746g_disco_ and _stm32f769i_disco_ boards.
 
If not, perhaps use `#ifndef CONFIG_DMA_STM32_V1` or similar to make it compile.

In case you choose to add this, here's my attempt at dts files for stm32f746i_disco. I can also make such patch for stm32f769g_disco if you like.

`boards/arm/stm32f756g_disco/stm32f756_disco.dts`:
```
&dma2 {
	status = "okay";
};

&quadspi {
	pinctrl-0 = <&quadspi_clk_pb2 &quadspi_bk1_ncs_pb6
		     &quadspi_bk1_io0_pd11 &quadspi_bk1_io1_pd12
		     &quadspi_bk1_io2_pe2 &quadspi_bk1_io3_pd13>;
	dmas = <&dma2 7 3 0x0000 0x03>;
	dma-names = "tx_rx";

	status = "okay";

	mx25r6435f: qspi-nor-flash@0 {
		compatible = "st,stm32-qspi-nor";
		label = "N25Q128A13EF840E";
		reg = <0>;
		spi-max-frequency = <108000000>;
		/* 128 Megabits = 16 Megabytes */
		size = <0x8000000>;
		status = "okay";
	};
};
```
`dts/arm/st/f7/stm32f7.dtsi`:
```
		quadspi: quadspi@a0001000 {
			compatible = "st,stm32-qspi";
			#address-cells = <1>;
			#size-cells = <0>;
			reg = <0xa0001000 0x34>;
			interrupts = <92 0>;
			clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00000002>;
			status = "disabled";
			label = "QUADSPI";
		};

```
(I can provide `.dts` files for QUADSPI support on these boards and theirs corresponding MCUs if it helps you.)

erwango replied:
@hjuul thanks for your interest. I'll try to add F7 support later this week.

erwango committed:
samples: spi_flash: Add compatibility with st,stm32-qspi-nor

Enable using st,stm32-qspi-nor compatible driver for this
sample.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

giansta commented:
We're running successfully the driver from PR #25806 on our board using STM32L462VE and QSPI flash Adesto AT25SL641, both in MCUboot and Zephyr application and we'd like to use this new one.
I made some tests, starting with the flash-shell. At the end the samples/drivers/flash_shell/ run successfully, I just tried few reads and some writes.
We have indeed to fix several build issues in our apps, including our version of flash_shell that now is only reading SFDP data, then failing to read the flash content (CS is not moved). I will try to work again on this in next days possibly. I should also try to build MCUboot.

My DTS looks like this:

```
&quadspi{
	pinctrl-0 = <&quadspi_bk1_ncs_pe11 &quadspi_clk_pe10
				 &quadspi_bk1_io0_pe12 &quadspi_bk1_io1_pb0
				 &quadspi_bk1_io2_pa7 &quadspi_bk1_io3_pa6>;
	dmas = <&dma1 5 5 0x0000 0x03>;
	dma-names = "tx_rx";

	status = "okay";

	at25sl641: qspi-nor-flash@0 {
		compatible = "st,stm32-qspi-nor";
		label = "AT25SL641";
		reg = <0>;
		spi-max-frequency = <133000000>;
		/* 64 Megabits = 8 Megabytes */
		size = <0x4000000>;
		jedec-id = [1f 43 17];
		status = "okay";
	};
};

&dma1 {
	status = "okay";
};
```

erwango commented:
Late bugs found. Under investigation

fzawadiak commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -0,0 +1,927 @@
+		LOG_ERR("Unexpected flash size: %u", flash_size);
+	}
+
+	LOG_INF("%s: %u MiBy flash", dev->name, (uint32_t)(flash_size >> 23));
```
Should it not be flash->size >> 20 ?

pabigot replied:
No.  Flash sizes for SPI NOR devices are specified in bits.  Long, boring story.

erwango replied:
Well, I agree this does not match with "%u MiBy flash"

pabigot replied:
Misled again by context: the division by 8 is a couple lines up in the original SPI NOR driver, so yes, it should be >> 20.

pabigot replied:
#31105, thanks for catching this @fzawadiak

erwango commented:
> Late bugs found. Under investigation

Issues solved. Pushed latest version including requested changes and fixes.

MaureenHelm commented on dts/bindings/qspi/st,stm32-qspi.yaml:
```diff
@@ -0,0 +1,52 @@
+
+include: base.yaml
+
+bus: qspi
```
Does the hardware support dual or octal modes? The NXP controller supports single, dual, quad, and octal modes, so I've been debating whether to use `bus: qspi` like this in #31034.

Does it make sense that all the vendors have completely different dts definitions of a qspi bus, without inheriting from a common binding? Nothing analogous to `spi-controller.yaml` and `spi-device.yaml`? @galak, @pabigot ?

erwango replied:
This hardware block is dual and quad only, not octal.
I have no strong opinion on common binding, I don't know if there is much to share, as even bus definition may differ.

MaureenHelm commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -0,0 +1,911 @@
+		.bus = DT_CLOCKS_CELL(STM32_QSPI_NODE, bus)
+	},
+	.irq_config = flash_stm32_qspi_irq_config_func,
+	.flash_size = DT_INST_PROP(0, size) / 8U,
+	.max_frequency = DT_INST_PROP(0, spi_max_frequency),
```
Can there ever be more than one instance?

erwango replied:
I don't think so, but from what I understand from @galak the `DT_INST_` version is the preferred one when possible.
At least I spared the iteration macro stuff.

MaureenHelm reviewed (commented):
> > The spi-nor sample and probably at least one flash subsystem sample should be updated to support a board (preferably B-L475E-IOT01A since that's what I've got coming...).
> 
> I agree that we need a way to test this driver in CI (or at least compile it). I think I can find a way to tweak `samples/drivers/flash_shell` to achieve this.
> Though, I'm not sure about the spi-nor sample, do you mean `samples/drivers/jesd216` ?

I found `samples/drivers/flash_shell` and `samples/subsys/fs/littlefs` to be useful while working on #31034.

fzawadiak commented:
FYI: I‘m working on H7 port, have it working with 4-4-4 mode and MDMA with cache invalidations. Dual mode is weird to do, as SFDP is read from both chips in parallel... So far MDMA is hacked in...

erwango committed:
samples/subsys: littlefs: Add disco_l475_iot1

Add disco_l475_iot1 as build platform for this sample

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

fzawadiak commented:
If you made dmas optional and used HAL_QSPI_Transmit_IT instead of HAL_QSPI_Transmit_DMA in case dmas are not defined then driver would work as is on H7. For deciding what communication mode to use we could maybe decide based on count of entries in pinctrl? At least between 1/4/dual 4 :-)

erwango committed:
drivers/flash: flash_stm32_qspi: Set DMA as optional

In order to ease reuse on other series, set DMA as optional
and use IT if no DMA channel is specified in the qspi node.

Tested on disco_l475_iot1

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@fzawadiak

> If you made dmas optional and used HAL_QSPI_Transmit_IT instead of HAL_QSPI_Transmit_DMA in case dmas are not defined then driver would work as is on H7. 

DMA is now optional in the driver. If "dmas" property is not set, qspi driver will work with IT.
@hjuul It should ease testing on F7.

> For deciding what communication mode to use we could maybe decide based on count of entries in pinctrl? At least between 1/4/dual 4 :-)

Let's discuss this as a follow up to this PR.

fzawadiak commented:
Super! Then as soon as it is merged I can add DTS and config for H7. Then will work on MDMA to add H7 DMA properly instead of hacking it in as I have it now.

fzawadiak commented on drivers/flash/Kconfig.stm32_qspi:
```diff
@@ -0,0 +1,21 @@
+	depends on SOC_FAMILY_STM32
+	default $(dt_compat_enabled,$(DT_COMPAT_ST_STM32_QSPI_NOR))
+	select USE_STM32_HAL_QSPI
+	select USE_STM32_HAL_DMA
```
Should it be automatic with optional DMAs? Currently DMA for H7 does not even compile...

erwango replied:
Txs for finding this. It is now made conditional on dts status.

erwango committed:
scripts/Kconfig: Add dt_node_has_prop Kconfig function

Add dt_node_has_prop function to query the presence of 'prop'
for given node label.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

pabigot reviewed (commented):
Finally getting to this.  Updating the jesd216 sample I get:
```
-- Found BOARD.dts: /mnt/nordic/zp/zephyr/boards/arm/disco_l475_iot1/disco_l475_iot1.dts
disco_l475_iot1.dts.pre.tmp:1815.42-1818.6: Warning (unit_address_format): /soc/quadspi@a0001000/qspi-nor-flash@0/partitions/partition@000D8000: unit name should not have leading 0s
disco_l475_iot1.dts.pre.tmp:165.29-174.5: Warning (spi_bus_bridge): /soc/quadspi@a0001000: node name for SPI buses should be 'spi'
  also defined at disco_l475_iot1.dts.pre.tmp:1793.10-1821.3
<stdout>: Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'
```

during build and at runtime

    Read SFDP not supported: device not JESD216-compliant (err -35)

which is `-ENOTSUP` even though `CONFIG_FLASH_JESD216_API=y`.

erwango commented:
> Finally getting to this. Updating the jesd216 sample I get:
> 
> ```
> -- Found BOARD.dts: /mnt/nordic/zp/zephyr/boards/arm/disco_l475_iot1/disco_l475_iot1.dts
> disco_l475_iot1.dts.pre.tmp:1815.42-1818.6: Warning (unit_address_format): /soc/quadspi@a0001000/qspi-nor-flash@0/partitions/partition@000D8000: unit name should not have leading 0s

This part is easy to fix

> disco_l475_iot1.dts.pre.tmp:165.29-174.5: Warning (spi_bus_bridge): /soc/quadspi@a0001000: node name for SPI buses should be 'spi'
>   also defined at disco_l475_iot1.dts.pre.tmp:1793.10-1821.3
> <stdout>: Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'
> ```

Yeah I've seen that but I'm a bit puzzled. Does it mean bus should be called spi instead of qspi ?

> 
> during build and at runtime
> 
> ```
> Read SFDP not supported: device not JESD216-compliant (err -35)
> ```
> 
> which is `-ENOTSUP` even though `CONFIG_FLASH_JESD216_API=y`.

Seems ok as FLASH_JESD216_API is not implemented on the driver. Is that mandatory ?

pabigot commented on dts/bindings/flash_controller/st,stm32-qspi-nor.yaml:
```diff
@@ -0,0 +1,35 @@
+properties:
+    reg:
+      required: true
+    spi-max-frequency:
```
This property name has to change, because [dtc infers](https://git.kernel.org/pub/scm/utils/dtc/dtc.git/tree/checks.c#n1090) bus type from the node name, or from property names in child nodes, so thinks this should be on a SPI bus, but it's not.  The Zephyr YAML `on-bus` stuff is not relevant to dtc.

Change it to `qspi-max-frequency` or `sck-frequency` or something that doesn't start with `spi`.

erwango replied:
Changed to `qspi-max-frequency`

fzawadiak replied:
Wouldn‘t max-frequency be even better? It is in qspi node?

pabigot commented:
> > disco_l475_iot1.dts.pre.tmp:165.29-174.5: Warning (spi_bus_bridge): /soc/quadspi@a0001000: node name for SPI buses should be 'spi'
> > also defined at disco_l475_iot1.dts.pre.tmp:1793.10-1821.3
> > : Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'
> 
> Yeah I've seen that but I'm a bit puzzled. Does it mean bus should be called spi instead of qspi ?

No.  Turns out it means that if dtc sees a property that begins with `spi-` then it thinks the node is a child of a spi bus, regardless of whether Zephyr's devicetree `on-bus` infrastructure claims otherwise.  So we need to change some property names.  Comment added.

> > which is `-ENOTSUP` even though `CONFIG_FLASH_JESD216_API=y`.
> 
> Seems ok as FLASH_JESD216_API is not implemented (yet) on the driver. Is that mandatory ?

Ah; no, it's fine for now, somehow I thought `FLASH_JESD216_API` was being selected by the board.  What's happening is the driver is selecting `FLASH_JESD216`--which is the sign that the API provided by the module enabled by that flag is available.  This means the sample, which requests `FLASH_JESD216_API` is building.  Leave that option unselected until the API is implemented.

erwango commented:
> > > disco_l475_iot1.dts.pre.tmp:165.29-174.5: Warning (spi_bus_bridge): /soc/quadspi@a0001000: node name for SPI buses should be 'spi'
> > > also defined at disco_l475_iot1.dts.pre.tmp:1793.10-1821.3
> > > : Warning (spi_bus_reg): Failed prerequisite 'spi_bus_bridge'
> > 
> > 
> > Yeah I've seen that but I'm a bit puzzled. Does it mean bus should be called spi instead of qspi ?
> 
> No. Turns out it means that if dtc sees a property that begins with `spi-` then it thinks the node is a child of a spi bus, regardless of whether Zephyr's devicetree `on-bus` infrastructure claims otherwise. So we need to change some property names. Comment added.
> 

Ok, that's easy to solve then. Txs for this because after several attempts, I gave up finding a reasonable explanation for this warning.

> > > which is `-ENOTSUP` even though `CONFIG_FLASH_JESD216_API=y`.
> > 
> > 
> > Seems ok as FLASH_JESD216_API is not implemented (yet) on the driver. Is that mandatory ?
> 
> Ah; no, it's fine for now, somehow I thought `FLASH_JESD216_API` was being selected by the board. What's happening is the driver is selecting `FLASH_JESD216`--which is the sign that the API provided by the module enabled by that flag is available. This means the sample, which requests `FLASH_JESD216_API` is building. Leave that option unselected until the API is implemented.

ok, will leave it untouched.

pabigot commented on dts/bindings/qspi/st,stm32-qspi.yaml:
```diff
@@ -0,0 +1,52 @@
+    dma-names:
+      required: true
+      description: |
+        DMA channel name. Expected value "tx_rx".
```
Can this be anything else?  When?  If it can't, maybe it should be `const` and not `required`.

Not blocking.

erwango replied:
I'm not following. Is the comment still valid in the light of the last commits?

pabigot replied:
Eh, well, the comment is here because I didn't *see* the last commits that changed it: I like to review one commit at a time.

I'm not requiring any changes, just pointing out something that confused me.  I'll take this as "too much hassle".

erwango replied:
> I like to review one commit at a time.

You're fully right. I usually don't like this kind of split in reviews neither.

pabigot commented on dts/bindings/qspi/st,stm32-qspi.yaml:
```diff
@@ -30,19 +30,17 @@ properties:
-      required: true
       description: |
-        DMA channel name. Expected value "tx_rx".
+        DMA channel name. If DMA should be used, expected value is "tx_rx".
```
Can this be refactored so it's not changing what was introduced in an earlier commit, or is that too much hassle?  ("It's too much hassle" is an acceptable answer.)

erwango replied:
What could be done is squash the latest commits that introduced **optional** dma.
I kept those separate in order not confusing on going reviews, but I can squash if required.

pabigot reviewed (commented):
Seems reasonable within the limits of my understanding and testing.  I do hope the full JESD216 API can be added soon; I think it'd make several things easier (e.g. managing 4-byte address support configuration; see early commits in #31045 that might get pulled out if #31042 gets merged but the Nordic stuff doesn't make progress).

erwango committed:
drivers/flash: stm32: Add qspi-nor flash-controller

Add support for DMA based STM32 QSPI NOR flash controller.
Driver configures both NOR flash and also QSPI hardware block.
Reuses existing jesd216 library.

QSPI hardware block handling is done through the use of Cube HAL API.
This requires the use of HAL interface also for DMA besides zephyr
DMA driver.
Zephyr DMA driver is used only for IRQ routing while HAL driver
handles the IP block. To achieve this it is required to:
-Configure both Cube and Zephyr drivers at init.
-Inform Zephyr driver that current channel handling will be done
by another instance and only a limited configuration should be done.
For this last part, a unused parameter is overridden in order to
transmit the information.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>
Signed-off-by: Piotr Mienkowski <piotr.mienkowski@gmail.com>

erwango commented:
Last update including following fixes :
-  fix bound for parameter headers cf https://github.com/zephyrproject-rtos/zephyr/pull/30371
- F7 compilation fixed (remains untested) https://github.com/zephyrproject-rtos/zephyr/pull/30864#discussion_r552616602

erwango committed:
dts/arm/st: wb: Add quadspi node

Add quadspi support to stm32wb series.
Tested on nucleo_wb55rg + nor flash.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Added stm32wb support

FRASTM reviewed (commented):
could the commit   cac002c7a013549fd7b1324d4f9fb65ba8e981fa be another PR ?

FRASTM commented on dts/arm/st/l4/stm32l4.dtsi:
```diff
@@ -171,6 +171,17 @@
+			compatible = "st,stm32-qspi";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			reg = <0xa0001000 0x34>;
```
reg = <0xa0001000 0x400>;
like other register banks --> 0xA0001000 - 0xA0001400

erwango replied:
Good catch, txs

erwango commented:
> could the commit cac002c be another PR ?

ok, right. Created https://github.com/zephyrproject-rtos/zephyr/pull/31303

erwango commented on drivers/flash/flash_stm32_qspi.c:
```diff
@@ -0,0 +1,923 @@
+
+#define STM32_QSPI_FIFO_THRESHOLD         8
+#define STM32_QSPI_CLOCK_PRESCALER_MAX  255
+#define STM32_DMA_HAL_OVERRIDE         0x7F
```
Added a new dedicated file.

fzawadiak commented:
DTS node for H7
```		quadspi: quadspi@52005000 {
			compatible = "st,stm32-qspi";
			#address-cells = <1>;
			#size-cells = <0>;
			reg = <0x52005000 0x400>;
			interrupts = <92 0>;
			clocks = <&rcc STM32_CLOCK_BUS_AHB3 0x00004000>;
			status = "disabled";
			label = "QUADSPI";
		};
```

erwango commented:
> DTS node for H7

Thanks. I had a test, this is working but I'm getting MDMA HAL warnings (even when not using DMA). If I manage to get rid of thoss, I'll include support in current PR. Otherwise this will have to wait a bit.

pabigot reviewed (approved):
Still want JESD216 API support but this is good and works for me on small devices.  Thanks.
